src/systolic_array/tensor.h

```diff
@@ -128,6 +129,10 @@ class TensorIndexIterator {
       state[i] = -halo[i].first;
   }
 
+  virtual TensorIndexIterator* clone() const {
```

Comment:
A copy constructor is preferable here, unless for some reason we don't want to allow that to happen.

Reply:
Yeah I tried to a copy ctor initially but C++ doesn't support virtual copy ctors, and the idiom to so is a virtual clone() function.

Reply:
Chatted offline. Return a unique_ptr here instead of a raw pointer.